pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -0,0 +1,260 @@
+                .insertUint64(startWeights[i].compress64(), _START_WEIGHT_OFFSET)
+                .insertUint32(endWeight.compress32(), _END_WEIGHT_OFFSET)
+                .insertUint5(uint256(18).sub(ERC20(address(token)).decimals()), _DECIMAL_DIFF_OFFSET);
+
+            normalizedSum = normalizedSum.add(endWeight);
+        }
+        // Ensure that the normalized weights sum to ONE
+        _require(normalizedSum == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
+
+        //_gradualUpdateTimestamps = newTimestamps
+        //    .insertUint32(startTime, _START_TIME_OFFSET)
+        //    .insertUint32(endTime, _END_TIME_OFFSET);
+
+        //emit GradualWeightUpdateScheduled(startTime, endTime, startWeights, endWeights);
+    }
+
+    function _computeScalingFactor(bytes32 tokenState) private view returns (uint256) {
+        uint256 decimalsDifference = tokenState.decodeUint5(_DECIMAL_DIFF_OFFSET);
+
+        return FixedPoint.ONE * 10**decimalsDifference;
+    }
+
+    // Private functions
+
+    /**
+     * @dev Returns a fixed-point number representing how far along the current weight change is, where 0 means the
+     * change has not yet started, and FixedPoint.ONE means it has fully completed.
+     */
+    function _calculateWeightChangeProgress() private view returns (uint256) {
+        /*uint256 currentTime = block.timestamp;
+        uint256 startTime = _gradualUpdateTimestamps.decodeUint32(_START_TIME_OFFSET);
+        uint256 endTime = _gradualUpdateTimestamps.decodeUint32(_END_TIME_OFFSET);
+
+        if (currentTime > endTime) {
+            return FixedPoint.ONE;
+        } else if (currentTime < startTime) {
+            return 0;
+        }
+
+        uint256 totalSeconds = endTime - startTime;
+        uint256 secondsElapsed = currentTime - startTime;
+
+        // In the degenerate case of a zero duration change, consider it completed (and avoid division by zero)
+        return totalSeconds == 0 ? FixedPoint.ONE : secondsElapsed.divDown(totalSeconds);*/
+
+        // For now, just return 0 (constant weights)
+        return 0;
+    }
+
+    function _interpolateWeight(bytes32 tokenData, uint256 pctProgress) private pure returns (uint256 finalWeight) {
+        uint256 startWeight = tokenData.decodeUint64(_START_WEIGHT_OFFSET).uncompress64();
+        uint256 endWeight = tokenData.decodeUint32(_END_WEIGHT_OFFSET).uncompress32();
+
+        if (pctProgress == 0 || startWeight == endWeight) return startWeight;
+        if (pctProgress >= FixedPoint.ONE) return endWeight;
+
+        if (startWeight > endWeight) {
+            uint256 weightDelta = pctProgress.mulDown(startWeight - endWeight);
+            return startWeight.sub(weightDelta);
+        } else {
+            uint256 weightDelta = pctProgress.mulDown(endWeight - startWeight);
+            return startWeight.add(weightDelta);
+        }
+    }
+
+    function _getValidTokenData(IERC20 token) private view returns (bytes32 tokenData) {
```

Comment:
```suggestion
    function _getTokenData(IERC20 token) private view returns (bytes32 tokenData) {
```

It feels weird to me to specify "valid" here. I understand this function enforces that we actually have data for `token` but I think that we can assume that functions return valid data (and revert otherwise) unless explicitly stated.